src/tactic/norm_cast.lean

```diff
@@ -24,269 +43,482 @@ try_for 1000 (mk_instance e)
+  new_y ← to_expr ``( @coe %%α %%β %%coe1 (@has_one.one %%α %%h2) ),
+  eq_y ← aux_squash y new_y,
+  let new_e := app (app op x) new_y,
+  pr ← mk_congr_arg (app op x) eq_y,
+  return ((), new_e, pr)
+)
 | _ := failed
-end
 
 /-
-This is the main heuristic used alongside the elim_cast and move_cast lemmas.
-An expression of the shape: op (↑(x : α) : γ) (↑(y : β) : γ)
-is rewritten as:            op (↑(↑(x : α) : β) : γ) (↑(y : β) : γ)
-when the squash_cast lemmas can prove that (↑(x : α) : γ) = (↑(↑(x : α) : β) : γ)
+assumption is used to discharge proofs in step 2
 -/
-private meta def heur (_ : unit) (e : expr) : tactic (unit × expr × expr) :=
-match e with
-| (app (expr.app op x) y) :=
-do
-    `(@coe %%α %%δ %%coe1 %%xx) ← return x,
-    `(@coe %%β %%γ %%coe2 %%yy) ← return y,
-    success_if_fail $ is_def_eq α β,
-    is_def_eq δ γ,
-
-    (do
-        coe3 ← mk_app `has_lift_t [α, β] >>= mk_instance',
-        new_x ← to_expr ``(@coe %%β %%δ %%coe2 (@coe %%α %%β %%coe3 %%xx)),
-        let new_e := app (app op new_x) y,
-        eq_x ← aux_simp x new_x,
-        pr ← mk_congr_arg op eq_x,
-        pr ← mk_congr_fun pr y,
-        return ((), new_e, pr)
-    ) <|> (do
-        coe3 ← mk_app `has_lift_t [β, α] >>= mk_instance',
-        new_y ← to_expr ``(@coe %%α %%δ %%coe1 (@coe %%β %%α %%coe3 %%yy)),
-        let new_e := app (app op x) new_y,
-        eq_y ← aux_simp y new_y,
-        pr ← mk_congr_arg (app op x) eq_y,
-        return ((), new_e, pr)
-    )
-| _ := failed
-end
+meta def prove : tactic unit := assumption
 
 /-
-simpa is used to discharge proofs
+TODO: norm_cast takes a list of expressions to use as lemmas for the discharger
 -/
-private meta def prove : tactic unit :=
-tactic.interactive.simpa none ff [] [] none
 
-private meta def post (s : simp_lemmas) (_ : unit) (e : expr) : tactic (unit × expr × expr) :=
-do
-    r ← mcond (is_prop e) (return `iff) (return `eq),
-    (new_e, pr) ← s.rewrite e prove r,
-    pr ← match r with
-    |`iff := mk_app `propext [pr]
-    | _   := return pr
-    end,
-    return ((), new_e, pr)
+/-
+This is an auxiliary function used in step 2.
+It tries to rewrite an expression using the elim_cast and move_cast lemmas.
+On failure, it calls the heuristic.
+-/
+meta def post (s : simp_lemmas) (_ : unit) (e : expr) : tactic (unit × expr × expr) :=
+( do
+  r ← mcond (is_prop e) (return `iff) (return `eq),
+  (new_e, pr) ← s.rewrite e prove r,
+  pr ← match r with
+  | `iff := mk_app `propext [pr]
+  | _    := return pr
+  end,
+  return ((), new_e, pr)
+) <|> heur () e
+
+/-
+The following auxiliary functions are used to handle numerals.
+-/
 
+-- prove ↑n = n where n is a numeral
+meta def aux_num_prove_eq (a b : expr) : tactic expr :=
+do
+  h ← to_expr ``(%%a = %%b),
+  s1 ← simp_lemmas.mk_default,
+  cache ← norm_cast_attr.get_cache,
+  let s := simp_lemmas.join s1 cache.down,
+  (_, pr) ← simplify s [] h,
+  some (_, tmp) ← expr.is_eq <$> infer_type pr,
+  is_def_eq tmp `(true),
```

Comment:
At the very least, this should be
```suggestion
  is_def_eq tmp `(true) reducible,
```
Off the top of my head, I'm not sure what the problem was with `simp_target`. `done` should have prevented it from any leftover metavars. But what you have now doesn't fix my concern, `is_def_eq` can still force computation. Restricting the unfolding to `reducible` should be fine.

Reply:
The problem was not with done, but the tactic failed when applying done, since simp_target was introducing goals. I'll try again to find a minimal example, but that was not at the top of my priority list with the classifier to add.